### PR TITLE
Keep file handle instead of entry in FileReader

### PR DESCRIPTION
**Pull request by MeAreJeenius**
# Description

When dealing with dosbox file IO (ie. DOS_{Open,Read,Seek,Close}File), it works with what dosbox calls file entries. These are the IDs which these functions will take in to find out which file to operate on.

Despite these sounding exactly like file handles, which is the assumption I was under when making FileReader, dosbox's internal system of open files uses a different set of IDs. I'm not sure why this is. There are functions to translate between the two.

From what I can tell, SC2K basically does something strange internally and invalidates the file handles somehow. When the entry was being used to keep track of the file, its status was not being updated. This means it was essentially a dangling handle, which is why it was reading out garbage. With these new changes having the handle being used instead, it seems that its corresponding entry is set to 0xff, which is used as the "null" ID, so it simply stops the rest of the batch file from being read, since the next read fails. This should bring the behavior to parity before this regression.

If I had to guess, the other problems mentioned in the linked issue are likely due to some deeper problem in how files are being handled.

## Related issues

Resolves the regression portion of #3518 

# Manual testing

Since the issue can't be generalized to a simple problem, the only thing that can be done is to test with the problematic batch file to ensure that the issue doesn't persist, and to test with a wide range of batch files to ensure that nothing is broken by the changes.

# Checklist

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [ ] commented on the particularly hard-to-understand areas of my code.
- [ ] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [ ] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

**Review by johnnovak — APPROVED** [review]
Thanks for the detailed explanation and the quick fix @MeAreJeenius, it all makes sense.

I think this is backport-worthy material.

**Comment by weirddan455** [comment]
Alternatively, can we just not keep a DOS file handle around at all?  Instead, in `GetFileReader` we could open the DOS file, read the entire thing into memory, close it, and never deal with the DOS interface again.  The state of the `FileReader` class can just be a `std::vector<uint8_t>` plus a position.  Should work fine baring very large files or something very strange like a self-modifying batch file.

Just an idea though.  What do you guys think?

**Comment by johnnovak** [comment]
Great pickup @weirddan455. I've looked into it and it turns out self-modifying MS-DOS batch files are a thing:

https://stackoverflow.com/a/27698558

The whole thread is worth reading, and multiple commenters confirm that remembering the file offset (not the line number) is the correct behaviour. So it seems to me the original code that got removed got this obscure detail right.

While it's not highly likely we'll run into self-modifying game installer batch files anytime soon, it's better to preserve the authentic behaviour, especially when it's not particularly hard to do so.

**Comment by FeralChild64** [comment]
We have a self-modifying .BAT file - Z:\AUTOEXEC.BAT.
At least it can get modified while being run if KEYB command is executed inside [autoexec] section to switch the code page.

**Comment by johnnovak** [comment]
Good writeup @MeAreJeenius. That hack is terrible, I vote for removal... Given that it doesn't even work as it's supposed to, I think that's a no-brainer. Then if the WINE folk need support for some similar functionality, we can treat that separately and implement it cleanly.

**Comment by weirddan455** [comment]
I just tested running a DOS program through Wine on my Fedora 39 system.  It uses Dosbox Staging 0.80.1 as the system Dosbox implementation.  There is no other fork available in the mainline repos.

This feature is broken here.  It runs with these arguments `arguments: dosbox -userconf -conf /home/daniel/.wine/dosdevices/c:/users/daniel/Temp/cfgd358.tmp`.  Staging goes through what looks like normal init, mounts some Wine paths as drive C and D, and then exits before even showing the window.  I'm not sure what's in that temporary config file because it gets deleted immediately afterwards.

For what it's worth I've been using Wine and its various forks for close to 20 years and I've never used this Dosbox integration.  I suppose the usecase is some Windows 95 program wanting to launch a DOS .exe.

I'm fine with removing this feature.  It sounds like a bad implementation and it doesn't work correctly anyway.

**Comment by johnnovak** [comment]
Great testing @weirddan455. We are all in agreement here @MeAreJeenius, that dirty hack has to go. Thanks for bringing the issue to our attention. 🏅

**Comment by johnnovak** [comment]
Nice one @MeAreJeenius. Are we all happy with this now @FeralChild64 and @weirddan455 ?

**Comment by FeralChild64** [comment]
Yes, dropping this hack is reasonable.

File: src/shell/file_reader.cpp
Comment (japsmits) [inline src/shell/file_reader.cpp:39]: Shouldn't this return an std::optional\<std::string\>? This way, it's easy to distinguish between a failed read (p.e. if the file doesn't open) and an empty line.

And maybe also a way to see what has failed. There are numerous ways I/O can fail, so I think we should be graceful around that subject.

Comment (johnnovak) [reply src/shell/file_reader.cpp]: Good point @japsmits.

**Comment by johnnovak** [comment]
> Looks good. I did some quick testing on my end to confirm the SC2000 bug is fixed and extended ASCII in batch files did not regress again.

@weirddan455 Given there have been further changes to the PR, would you be able to quickly re-execute your test cases before I press the merge button? Thanks mate.

**Comment by weirddan455** [comment]
> @weirddan455 Given there have been further changes to the PR, would you be able to quickly re-execute your test cases before I press the merge button? Thanks mate.

Yep, tests are still good.

**Comment by johnnovak** [comment]
Actually, we can't backport this as it was built on top of the major shell refactor stuff. Tried to cherry pick but I'm getting tons of conflicts. But it's ok, the bug it fixes is very obscure.

**Comment by weirddan455** [comment]
@Torinde That feature request is not really related to Wine integration.  On the Wine side, they're expecting to launch our GUI not have us interact with the terminal.

We've already decided both of those features are out of scope for Staging.  Wine integration is one of the rare cases where we've broken compatibility with original DOSBox.  We must have accidentally broke it some time before this PR and then intentionally removed a hack for it in this PR.  It was a rarely used and poorly implemented feature from the start so I don't feel too bad about it.

The render to console stuff is unlikely to ever happen either.  There's a lot of nuance to getting something like that implemented and it doesn't really help with playing games (aside from text mode games I guess but that's pretty niche and those games will still work fine in our GUI).

**Comment by johnnovak** [comment]
Yep fully agree with @weirddan455 on all points.

**Pull request merged**
